src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/PolicyFunctions.kt

```diff
@@ -25,42 +32,77 @@ fun <T : ILicenceCondition> conditionChanges(current: List<T>, other: List<T>) =
   amendedConditions(current, other)
 )
 
-/*
- *  For each condition, create List item of custom type
- *  Add new list of HashMap<placeholder, variable> to condition list
- *  Iterate through placeholders and add to HashMap with corresponding date field ({} from tpl)
- */
-fun <T : ILicenceCondition> getPolicyPlaceholders(conditions: List<T>): HashMap<String, HashMap<String, String>> {
-  val placeholderPattern = Regex("(?<=\\[)(.*?)(?=])")
-  val variablePattern = Regex("(?<=\\{)(.*?)(?=})")
-  val results = HashMap<String, HashMap<String, String>>()
+fun <T : Any> licencePolicyChanges(
+  presentConditions: List<AdditionalCondition>,
+  previousConditions: List<IAdditionalCondition<T>>,
+  currentConditions: List<IAdditionalCondition<T>>,
+  allReplacements: List<Replacements>
+): List<LicenceConditionChanges> {
+  val conditionsToCheck = currentConditions.filter { current ->
+    presentConditions.any { present -> current.code == present.code }
+  }
+  val changes = conditionChanges(conditionsToCheck, previousConditions, allReplacements)
 
-  conditions.filter { it.requiresInput }.forEach {
-    val placeholders = placeholderPattern.findAll(it.text).iterator()
-    val variables = variablePattern.findAll(it.tpl ?: "").toSet().iterator()
-    val keyValuePairs = HashMap<String, String>()
+  return changes.map { change -> change.copy(sequence = presentConditions.find { condition -> condition.code == change.code }?.sequence) }
+}
 
-    while (placeholders.hasNext() && variables.hasNext()) {
-      keyValuePairs[placeholders.next().value] = variables.next().value
-    }
+fun <T : Any> conditionChanges(
+  previousConditions: List<IAdditionalCondition<T>>,
+  currentConditions: List<IAdditionalCondition<T>>,
+  allReplacements: List<Replacements>
+): List<LicenceConditionChanges> = previousConditions.mapNotNull { previous ->
+  val current = currentConditions.find { it.code == previous.code }
+  if (current != null)
+    conditionUpdated(previous, current)
+  else
+    conditionRemoved(previous, allReplacements)
+}
 
-    results[it.code] = keyValuePairs
+private fun <T : Any> conditionUpdated(
+  previous: IAdditionalCondition<T>,
+  current: IAdditionalCondition<T>,
+): LicenceConditionChanges? {
+  val textHasChanged = previous.text != current.text
+  val removed = (previous.inputs ?: emptyList()) - (current.inputs ?: emptyList()).toSet()
+  val added = (current.inputs ?: emptyList()) - (previous.inputs ?: emptyList()).toSet()
```

Comment:
just working out the difference between the two lists. Updates will appear in both lists - "You removed this variant of condition A and added this other variant of condition A"

This isn't used by the frontend but returning as it help with debugging and seeing why things are different